src/databased/runtime.rs

```diff
@@ -224,6 +225,34 @@ impl Runtime {
                 self.database.set_offer_status(&offer, &status)?;
             }
 
+            CtlMsg::CleanDanglingOffers => {
+                let checkpointed_pub_offers: Vec<PublicOffer> = self
+                    .database
+                    .get_all_checkpoint_info()?
+                    .iter()
+                    .filter_map(|(_, info)| {
+                        CheckpointEntry::strict_decode(std::io::Cursor::new(info)).ok()
+                    })
+                    .map(|c| c.public_offer)
+                    .collect();
+                let dangling_offers: Vec<PublicOffer> = self
+                    .database
+                    .get_offers(OfferStatusSelector::InProgress)?
+                    .drain(..)
+                    .filter_map(|o| {
+                        if !checkpointed_pub_offers.contains(&o.offer) {
+                            Some(o.offer)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                for offer in dangling_offers.iter() {
```

Comment:
Style: i think we can avoid collect and creating a new iterator and just have a for_each after the filter_map. Or is it like that because self.database?

Reply:
> Or is it like that because self.database?

It is like that so we can report the Error of `set_offer_status` to the upper level. Is there a better way of doing that?

Reply:
Not sure if it is better but we can achieve the same with:
```rust
                self.database
                    .get_offers(OfferStatusSelector::InProgress)?
                    .drain(..)
                    .filter_map(|o| {
                        if !checkpointed_pub_offers.contains(&o.offer) {
                            Some(o.offer)
                        } else {
                            None
                        }
                    })
                    .map(|offer| {
                        self.database
                            .set_offer_status(&offer, &OfferStatus::Ended(Outcome::Abort))
                    })
                    .collect::<Result<_, _>>()?;
```
Collect can transform a `Vec<Res<>>` into `Res<Vec<>>`, this should fail on the first error.

Reply:
I found this as well in the meantime from rust by example: https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect

I think this is indeed preferable, will apply this pattern elsewhere when writing new code.

Reply:
635aa9af9badaa1b466067f24d568e4d8d4d9b72